Model/Validate.php

```diff
@@ -87,9 +95,19 @@ public function beforeSave()
     {
         $postData = $this->request->getPostValue();
         $postValues = $postData['groups']['general']['fields'];
-        $shopId = $postValues['shop_id']['value'];
-        $shopPassword = $postValues['shop_password']['value'];
-        $smartCheck = $postValues['smart_check']['value'];
+        $storeId = $this->request->getParam('store', 0);
+
+        if (isset($postValues['shop_id']['value'])) {
+            $shopId = $postValues['shop_id']['value'];
+        } elseif (isset($postValues['shop_id']['inherit']) && $postValues['shop_id']['inherit'] == 1) {
+            $shopId = $this->dataHelper->getShopId($storeId);
+        }
+
+        if (isset($postValues['shop_password']['value'])) {
+            $shopPassword = $postValues['shop_password']['value'];
+        } elseif (isset($postValues['shop_password']['inherit']) && $postValues['shop_password']['inherit'] == 1) {
+            $shopPassword = $this->dataHelper->getShopPw($storeId);
+        }
```

Comment:
The above code is somewhat duplicate. You can make the above code a function and may be pass an index to check
```
$valueToReturn = '';
if (isset($postValues[$index]['value'])) {
    $valueToReturn = $postValues[$index]['value'];
} elseif (isset($postValues[$index]['inherit']) && $postValues[$index]['inherit'] == 1) {
    $valueToReturn = $this->getValue($index, $storeId); //here getValue can call the functions on the basis of index given
}
```
Check if something like above is appropriate for u